packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -326,9 +360,16 @@ export class ExecutionEngine {
   }
 
   /**
-   * Wait for the next block to  be processed on-chain
+   * Wait for the next block to  be processed on-chain, monitoring transactions
+   * for timeout at the same time. If timing out transactions complete
+   * the batch we return before the next block.
    */
-  private async _newBlock(state: ExecutionEngineState): Promise<void> {
+  private async _newBlockOrTimeout(
```

Comment:
I think we can wait for a new block, and run the timeout logic after it gets mined. If a future timeouts between two blocks it'd wait at most the block interval time of the chain, which is in the order of seconds.

Reply:
This will move timeout handling to process future tick, making things simpler / more consistent.